File: sass-api/internal/services/user_service.go
package services

import (
	"errors"
	"time"

	"github.com/Aebroyx/sass-api/internal/config"
	"github.com/Aebroyx/sass-api/internal/domain/models"
	"github.com/Aebroyx/sass-api/internal/pagination"
	"github.com/golang-jwt/jwt/v5"

	"golang.org/x/crypto/bcrypt"
	"gorm.io/gorm"
)

type UserService struct {
	db           *gorm.DB
	config       *config.Config
	tokenService *TokenService
}

// UserQueryParams represents the query parameters for user listing
type UserQueryParams struct {
	Page     int    `json:"page" form:"page" binding:"min=1"`
	PageSize int    `json:"pageSize" form:"pageSize" binding:"min=1,max=100"`
	Search   string `json:"search" form:"search"`
	RoleID   uint   `json:"role_id" form:"role_id"`
	SortBy   string `json:"sortBy" form:"sortBy" binding:"omitempty,oneof=name email role_id created_at"`
	SortDesc bool   `json:"sortDesc" form:"sortDesc"`
}

// UserListResponse represents the paginated response for user listing
type UserListResponse struct {
	Data       []models.Users `json:"data"`
	Total      int64          `json:"total"`
	Page       int            `json:"page"`
	PageSize   int            `json:"pageSize"`
	TotalPages int            `json:"totalPages"`
}

func NewUserService(db *gorm.DB, config *config.Config, tokenService *TokenService) *UserService {
	return &UserService{
		db:           db,
		config:       config,
		tokenService: tokenService,
	}
}

// Register creates a new user with the provided registration data
func (s *UserService) Register(req *models.RegisterRequest) (*models.RegisterResponse, error) {
	// Check if username already exists
	var existingUser models.Users
	if err := s.db.Where("username = ?", req.Username).First(&existingUser).Error; err == nil {
		return nil, errors.New("username already exists")
	} else if !errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, err
	}

	// Check if email already exists
	if err := s.db.Where("email = ?", req.Email).First(&existingUser).Error; err == nil {
		return nil, errors.New("email already exists")
	} else if !errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, err
	}

	// Hash password
	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
	if err != nil {
		return nil, err
	}

	// Get default role
	var defaultRole models.Role
	if err := s.db.Where("is_default = ?", true).First(&defaultRole).Error; err != nil {
		// Fallback to user role if no default
		if err := s.db.Where("name = ?", "user").First(&defaultRole).Error; err != nil {
			return nil, errors.New("default role not found")
		}
	}

	// Create new user with RoleID
	user := models.Users{
		Username: req.Username,
		Email:    req.Email,
		Password: string(hashedPassword),
		Name:     req.Name,
		RoleID:   defaultRole.ID,
	}

	if err := s.db.Create(&user).Error; err != nil {
		return nil, err
	}

	// Preload role for response
	s.db.Preload("Role").First(&user, user.ID)

	// Return user data without password
	return &models.RegisterResponse{
		ID:       user.ID,
		Username: user.Username,
		Email:    user.Email,
		Name:     user.Name,
		IsActive: user.IsActive,
		Role: models.RoleResponse{
			ID:          user.Role.ID,
			Name:        user.Role.Name,
			DisplayName: user.Role.DisplayName,
			Description: user.Role.Description,
			IsDefault:   user.Role.IsDefault,
			IsActive:    user.Role.IsActive,
			CreatedAt:   user.Role.CreatedAt,
			UpdatedAt:   user.Role.UpdatedAt,
		},
	}, nil
}

// Login authenticates a user and returns tokens
// Deprecated: Use LoginWithContext instead
func (s *UserService) Login(req *models.LoginRequest) (*models.LoginResponse, error) {
	return s.LoginWithContext(req, "", "")
}

// LoginWithContext authenticates a user and returns tokens with IP and UserAgent tracking
func (s *UserService) LoginWithContext(req *models.LoginRequest, ipAddress, userAgent string) (*models.LoginResponse, error) {
	// Find user by username with role preloaded
	var user models.Users
	if err := s.db.Preload("Role").Where("username = ?", req.Username).First(&user).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, errors.New("invalid username or password")
		}
		return nil, err
	}

	// Verify password
	if err := bcrypt.CompareHashAndPassword([]byte(user.Password), []byte(req.Password)); err != nil {
		return nil, errors.New("invalid username or password")
	}

	// Check if user is active
	if !user.IsActive {
		return nil, errors.New("user is not active")
	}

	// Generate access token (JWT)
	accessToken, accessExp, err := s.generateToken(user, s.config.JWTExpiry)
	if err != nil {
		return nil, err
	}

	// Create refresh token in database (if tokenService is available)
	var refreshTokenStr string
	if s.tokenService != nil {
		refreshToken, err := s.tokenService.CreateRefreshToken(user.ID, ipAddress, userAgent)
		if err != nil {
			return nil, err
		}
		refreshTokenStr = refreshToken.Token
	} else {
		// Fallback to JWT-based refresh token for backward compatibility
		refreshTokenStr, _, err = s.generateToken(user, s.config.RefreshTokenExpiry)
		if err != nil {
			return nil, err
		}
	}

	// Create response
	return &models.LoginResponse{
		User: models.RegisterResponse{
			ID:       user.ID,
			Username: user.Username,
			Email:    user.Email,
			Name:     user.Name,
			IsActive: user.IsActive,
			Role: models.RoleResponse{
				ID:          user.Role.ID,
				Name:        user.Role.Name,
				DisplayName: user.Role.DisplayName,
				Description: user.Role.Description,
				IsDefault:   user.Role.IsDefault,
				IsActive:    user.Role.IsActive,
				CreatedAt:   user.Role.CreatedAt,
				UpdatedAt:   user.Role.UpdatedAt,
			},
		},
		Token: models.TokenResponse{
			AccessToken:  accessToken,
			RefreshToken: refreshTokenStr,
			TokenType:    "Bearer",
			ExpiresIn:    int64(time.Until(accessExp).Seconds()),
		},
	}, nil
}

// generateToken generates a JWT token for the user
func (s *UserService) generateToken(user models.Users, expiry time.Duration) (string, time.Time, error) {
	expirationTime := time.Now().Add(expiry)
	claims := &models.Claims{
		UserID:   user.ID,
		Username: user.Username,
		Email:    user.Email,
		RoleID:   user.RoleID,
		RoleName: user.Role.Name,
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(expirationTime),
			IssuedAt:  jwt.NewNumericDate(time.Now()),
			NotBefore: jwt.NewNumericDate(time.Now()),
			Issuer:    "the-blade-api",
			Subject:   user.Username,
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	tokenString, err := token.SignedString([]byte(s.config.JWTSecret))
	if err != nil {
		return "", time.Time{}, err
	}

	return tokenString, expirationTime, nil
}

// GetAllUsers retrieves users with pagination, search, and filters
func (s *UserService) GetAllUsers(params pagination.QueryParams) (*pagination.PaginatedResponse, error) {
	config := pagination.PaginationConfig{
		Model: &models.Users{},
		BaseCondition: map[string]interface{}{
			"is_deleted": false,
		},
		SearchFields: []string{"name", "email", "username"},
		FilterFields: map[string]string{
			"role_id":    "role_id",
			"name":       "name",
			"email":      "email",
			"username":   "username",
			"created_at": "created_at",
			"updated_at": "updated_at",
		},
		DateFields: map[string]pagination.DateField{
			"created_at": {
				Start: "created_at",
				End:   "created_at",
			},
			"updated_at": {
				Start: "updated_at",
				End:   "updated_at",
			},
		},
		SortFields: []string{
			"name",
			"email",
			"role_id",
			"created_at",
			"updated_at",
		},
		DefaultSort:  "created_at",
		DefaultOrder: "DESC",
		Relations:    []string{"Role"},
	}

	paginator := pagination.NewPaginator(s.db)
	return paginator.Paginate(params, config)
}

func (s *UserService) GetUserById(id string) (models.Users, error) {
	var user models.Users
	if err := s.db.Preload("Role").Where("id = ?", id).First(&user).Error; err != nil {
		return models.Users{}, err
	}
	return user, nil
}

// CreateUser creates a new user with the provided data
func (s *UserService) CreateUser(req *models.CreateUserRequest) (*models.CreateUserResponse, error) {
	// Check if username already exists
	var existingUser models.Users
	if err := s.db.Where("username = ?", req.Username).First(&existingUser).Error; err == nil {
		return nil, errors.New("username already exists")
	} else if !errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, err
	}

	// Check if email already exists
	if err := s.db.Where("email = ?", req.Email).First(&existingUser).Error; err == nil {
		return nil, errors.New("email already exists")
	} else if !errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, err
	}

	// Validate role exists
	var role models.Role
	if err := s.db.First(&role, req.RoleID).Error; err != nil {
		return nil, errors.New("invalid role")
	}

	// Hash password
	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
	if err != nil {
		return nil, err
	}

	// Validate IsActive is provided
	if req.IsActive == nil {
		return nil, errors.New("is_active is required")
	}

	// Create new user
	user := models.Users{
		Username: req.Username,
		Email:    req.Email,
		Password: string(hashedPassword),
		Name:     req.Name,
		RoleID:   req.RoleID,
		IsActive: *req.IsActive,
	}

	if err := s.db.Create(&user).Error; err != nil {
		return nil, err
	}

	// Preload role for response
	s.db.Preload("Role").First(&user, user.ID)

	// Return user data without password
	return &models.CreateUserResponse{
		ID:       user.ID,
		Username: user.Username,
		Email:    user.Email,
		Name:     user.Name,
		IsActive: user.IsActive,
		Role: models.RoleResponse{
			ID:          user.Role.ID,
			Name:        user.Role.Name,
			DisplayName: user.Role.DisplayName,
			Description: user.Role.Description,
			IsDefault:   user.Role.IsDefault,
			IsActive:    user.Role.IsActive,
			CreatedAt:   user.Role.CreatedAt,
			UpdatedAt:   user.Role.UpdatedAt,
		},
		CreatedAt: user.CreatedAt,
	}, nil
}

func (s *UserService) UpdateUser(id string, req *models.UpdateUserRequest) (*models.Users, error) {
	var user models.Users
	if err := s.db.Preload("Role").Where("id = ?", id).First(&user).Error; err != nil {
		return nil, err
	}

	// Validate role exists if being changed
	if req.RoleID != user.RoleID {
		var role models.Role
		if err := s.db.First(&role, req.RoleID).Error; err != nil {
			return nil, errors.New("invalid role")
		}
	}

	// Validate IsActive is provided
	if req.IsActive == nil {
		return nil, errors.New("is_active is required")
	}

	// Update user fields
	user.Username = req.Username
	user.Email = req.Email
	user.Name = req.Name
	user.RoleID = req.RoleID
	user.IsActive = *req.IsActive

	// Only update password if provided
	if req.Password != "" {
		hashedPassword, err := bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
		if err != nil {
			return nil, err
		}
		user.Password = string(hashedPassword)
	}

	// Update user - use Select to explicitly specify fields to update (including is_active even when false)
	fieldsToUpdate := []string{"username", "email", "name", "role_id", "is_active"}
	if req.Password != "" {
		fieldsToUpdate = append(fieldsToUpdate, "password")
	}

	if err := s.db.Model(&user).Select(fieldsToUpdate).Updates(&user).Error; err != nil {
		return nil, err
	}

	// Reload with role
	s.db.Preload("Role").First(&user, user.ID)

	return &user, nil
}

func (s *UserService) DeleteUser(id string) (*models.Users, error) {
	var user models.Users
	if err := s.db.Transaction(func(tx *gorm.DB) error {
		// First, get the user to verify it exists
		if err := tx.Where("id = ?", id).First(&user).Error; err != nil {
			return err
		}

		// Prevent deletion of root user only
		if user.Username == "root" || user.Email == "root@localhost" {
			return errors.New("cannot delete root user")
		}

		// Mark as deleted (in addition to Gorm's soft delete)
		if err := tx.Model(&models.Users{}).Where("id = ?", id).Updates(map[string]interface{}{
			"is_deleted": true,
		}).Error; err != nil {
			return err
		}

		// Soft delete the user (sets deleted_at)
		if err := tx.Where("id = ?", id).Delete(&user).Error; err != nil {
			return err
		}

		return nil
	}); err != nil {
		return nil, err
	}

	return &user, nil
}

func (s *UserService) ResetUserPassword(id string, req *models.ResetUserPasswordRequest) (*models.Users, error) {
	var user models.Users
	if err := s.db.Where("id = ?", id).First(&user).Error; err != nil {
		return nil, err
	}

	// Verify current password
	if err := bcrypt.CompareHashAndPassword([]byte(user.Password), []byte(req.CurrentPassword)); err != nil {
		return nil, errors.New("invalid current password")
	}

	// Verify new password and confirm password
	if req.NewPassword != req.ConfirmPassword {
		return nil, errors.New("new password and confirm password do not match")
	}

	// Hash new password
	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(req.NewPassword), bcrypt.DefaultCost)
	if err != nil {
		return nil, err
	}

	// Update user password
	user.Password = string(hashedPassword)

	// Update user
	if err := s.db.Model(&user).Updates(&user).Error; err != nil {
		return nil, err
	}

	// Reload with role
	s.db.Preload("Role").First(&user, user.ID)

	return &user, nil
}

File: sass-ui/src/services/userService.ts
import { User } from '@/store/features/authSlice';
import axiosInstance, { handleApiError } from '@/lib/axios';
import { FilterCondition } from '@/components/modals/AdvancedFilterModal';

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8080/api';

interface RegisterRequest {
  username: string;
  email: string;
  password: string;
  name: string;
}

interface RegisterResponse {
  id: number;
  username: string;
  email: string;
  name: string;
  role: string;
}

interface LoginRequest {
  username: string;
  password: string;
}

interface TokenResponse {
  access_token: string;
  refresh_token: string;
  token_type: string;
  expires_in: number;
}

interface LoginResponse {
  user: RegisterResponse;
}

interface RoleResponse {
  id: number;
  name: string;
  display_name: string;
  description: string;
  is_default: boolean;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

interface GetUserResponse {
  id: number;
  username: string;
  email: string;
  name: string;
  role: RoleResponse;
  created_at: string;
  updated_at: string;
  is_active: boolean;
}

interface CreateUserRequest {
  name: string;
  email: string;
  username: string;
  password: string;
  role_id: number;
  is_active: boolean;
}

interface ErrorResponse {
  status: 'error';
  message: string;
  code?: string;
  details?: unknown;
}

interface UpdateUserRequest {
  name: string;
  email: string;
  role_id: number;
  username: string;
  password: string;
  is_active: boolean;
}

interface ResetPasswordRequest {
  current_password: string;
  new_password: string;
  confirm_password: string;
}

// Response represents the standard API response structure
interface ApiResponse<T> {
  status: string;
  message: string;
  data: T;
}

// PaginatedResponse represents the pagination response structure
export interface PaginatedResponse<T> {
  data: T[];
  total: number;
  page: number;
  pageSize: number;
  totalPages: number;
}

export interface GetAllUsersParams {
  page: number;
  pageSize: number;
  search?: string;
  sortBy?: string;
  sortDesc?: boolean;
  filters?: FilterCondition[];
}

class UserService {
  private async fetchWithError(url: string, options: RequestInit = {}) {
    console.log('Making request to:', url, 'with options:', {
      ...options,
      body: options.body ? JSON.parse(options.body as string) : undefined
    });

    const response = await fetch(url, {
      ...options,
      headers: {
        'Content-Type': 'application/json',
        ...options.headers,
      },
    });
    const responseText = await response.text();

    if (!response.ok) {
      let errorResponse: ErrorResponse;
      try {
        errorResponse = JSON.parse(responseText);
      } catch (e) {
        console.error('Error parsing error response:', e);
        throw new Error('Something went wrong');
      }
      throw new Error(JSON.stringify(errorResponse));
    }

    try {
      return JSON.parse(responseText);
    } catch (e) {
      console.error('Error parsing response:', e);
      throw new Error('Invalid response from server');
    }
  }

  async register(data: RegisterRequest): Promise<RegisterResponse> {
    return this.fetchWithError(`${API_URL}/auth/register`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  async login(data: LoginRequest): Promise<LoginResponse> {
    return this.fetchWithError(`${API_URL}/auth/login`, {
      method: 'POST',
      body: JSON.stringify(data),
      credentials: 'include',
    });
  }

  async logout(): Promise<void> {
    // You might want to call a logout endpoint here if you have one
    // For now, we'll just clear the tokens on the client side
  }

  async getCurrentUser(token: string): Promise<User> {
    return this.fetchWithError(`${API_URL}/me`, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });
  }

  // Add new user management methods
  async getAllUsers(params: GetAllUsersParams): Promise<PaginatedResponse<GetUserResponse>> {
    try {
      const queryParams = new URLSearchParams();
      if (params.page) queryParams.append('page', params.page.toString());
      if (params.pageSize) queryParams.append('pageSize', params.pageSize.toString());
      if (params.search) queryParams.append('search', params.search);
      if (params.sortBy) queryParams.append('sortBy', params.sortBy);
      if (params.sortDesc !== undefined) queryParams.append('sortDesc', params.sortDesc.toString());
      
      // Handle advanced filter conditions
      if (params.filters && params.filters.length > 0) {
        // Send filters as JSON string in query params
        queryParams.append('filters', JSON.stringify(params.filters));
      }

      const response = await axiosInstance.get<ApiResponse<PaginatedResponse<GetUserResponse>>>(`/users?${queryParams.toString()}`);
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  }

  async getUserById(id: string | number): Promise<GetUserResponse> {
    try {
      const response = await axiosInstance.get<{ data: GetUserResponse }>(`/user/${id}`);
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  }

  async createUser(data: CreateUserRequest): Promise<GetUserResponse> {
    try {
      const response = await axiosInstance.post<GetUserResponse>('/user/create', data);
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  }

  async updateUser(id: string | number, data: UpdateUserRequest): Promise<GetUserResponse> {
    try {
      const response = await axiosInstance.put<GetUserResponse>(`/user/${id}`, data);
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  }

  async deleteUser(id: string | number): Promise<void> {
    try {
      await axiosInstance.delete(`/user/${id}`);
    } catch (error) {
      throw handleApiError(error);
    }
  }

  async softDeleteUser(id: string | number): Promise<void> {
    try {
      await axiosInstance.put(`/user/${id}/soft-delete`);
    } catch (error) {
      throw handleApiError(error);
    }
  }

  async resetPassword(id: string | number, data: ResetPasswordRequest): Promise<GetUserResponse> {
    try {
      const response = await axiosInstance.post<ApiResponse<GetUserResponse>>(`/user/reset-password/${id}`, data);
      return response.data.data;
    } catch (error) {
      throw handleApiError(error);
    }
  }
  
}

export const userService = new UserService();
export type { GetUserResponse, RegisterRequest, RegisterResponse, LoginRequest, TokenResponse, LoginResponse, ResetPasswordRequest };
